src/Converter/Feature2Texture.js

```diff
@@ -37,66 +38,81 @@ function _drawPolygon(ctx, vertices, indices, style, size, extent, invCtxScale,
 
     // draw line or edge of polygon
     if (style.stroke.color) {
-        strokeStyle(style, ctx, invCtxScale);
+        strokeStyle(style, ctx, invCtxScale, featureCtx);
         ctx.stroke();
     }
 
     // fill polygon only
     if (canBeFilled && (style.fill.color || style.fill.pattern)) {
-        fillStyle(style, ctx, invCtxScale);
+        fillStyle(style, ctx, invCtxScale, featureCtx);
         ctx.fill();
     }
 }
 
-function fillStyle(style, ctx, invCtxScale) {
+function fillStyle(style, ctx, invCtxScale, featureCtx) {
     if (style.fill.pattern && ctx.fillStyle.src !== style.fill.pattern.src) {
         ctx.fillStyle = ctx.createPattern(style.fill.pattern, 'repeat');
         if (ctx.fillStyle.setTransform) {
             ctx.fillStyle.setTransform(matrix.scale(invCtxScale));
         } else {
             console.warn('Raster pattern isn\'t completely supported on Ie and edge');
         }
-    } else if (style.fill.color && ctx.fillStyle !== style.fill.color) {
-        ctx.fillStyle = style.fill.color;
+    } else if (style.fill.color) {
+        const fillStyle = readExpression(style.fill.color, featureCtx);
+        if (ctx.fillStyle !== fillStyle) {
+            ctx.fillStyle = fillStyle;
+        }
     }
-    if (style.fill.opacity !== ctx.globalAlpha) {
-        ctx.globalAlpha = style.fill.opacity;
+
+    const opacity = readExpression(style.fill.opacity, featureCtx);
+    if (opacity !== ctx.globalAlpha) {
+        ctx.globalAlpha = opacity;
     }
 }
 
-function strokeStyle(style, ctx, invCtxScale) {
-    if (ctx.strokeStyle !== style.stroke.color) {
-        ctx.strokeStyle = style.stroke.color;
+function strokeStyle(style, ctx, invCtxScale, featureCtx) {
+    const strokeStyle = readExpression(style.stroke.color, featureCtx);
+    if (ctx.strokeStyle !== strokeStyle) {
+        ctx.strokeStyle = strokeStyle;
     }
-    const width = (style.stroke.width || 2.0) * invCtxScale;
+    const width = (readExpression(style.stroke.width, featureCtx) || 2.0) * invCtxScale;
     if (ctx.lineWidth !== width) {
         ctx.lineWidth = width;
     }
-    const alpha = style.stroke.opacity == undefined ? 1.0 : style.stroke.opacity;
+    const opacity = readExpression(style.stroke.opacity, featureCtx);
+    const alpha = opacity == undefined ? 1.0 : opacity;
     if (alpha !== ctx.globalAlpha && typeof alpha == 'number') {
         ctx.globalAlpha = alpha;
     }
-    if (ctx.lineCap !== style.stroke.lineCap) {
-        ctx.lineCap = style.stroke.lineCap;
+    const lineCap = readExpression(style.stroke.lineCap, featureCtx);
+    if (ctx.lineCap !== lineCap) {
+        ctx.lineCap = lineCap;
```

Comment:
I might be missing some point, but wouldn't just doing `ctx.lineCap = readExpression(style.stroke.lineCap, featureCtx);` bring the same result, and prevent from checking difference each time ?

Reply:
The canvas context change  (`ctx`) influences on the performance. We check to avoid some changes.